### ISSUE TITLE
[PROPOSAL] Code quality analysis

### ISSUE THREAD

**Issue by zethuman** [opensearch-project/opensearch-go#307]
### What are you proposing?
_Analyze overall code quality in the automated pipeline._

### What problems are you trying to solve?
_When make PR pull , a contributors wants to check up requested code, so they can fix abstract errors and cover with tests_

#### Are there any security considerations? 
_I haven't research deeper. If there is some cases with security, let's discuss it_

#### Are there any breaking changes to the API
_No, that is only loyality for developers_

### Why should it be built? Any reason not to?
_Describe the value that this feature will bring to the OpenSearch community, as well as what impact it has if it isn't built, or new risks if it is. Highlight opportunities for additional research._

### What will it take to execute?
_We should integrate some static code quality analyzer like a SonarQube_

**Comment by dblock** [comment]
+1 on any kind of static analysis tooling

**Comment by Jakob3xD** [comment]
I have a few questions related to that.
- How do we define code quality?
- Do we only use the defaults from the action provider or do we define them by our self?
- If so, how would we define them by our self?
- What do we do if some defaults don't match our wanted code quality or lets say we don't like some of the defaults?

**Comment by Jakob3xD** [comment]
Does it scans the PR code quality or the whole repo?

I am generally okay with adding it and see how it works, but I don't think it makes sense in current state of this repo. As we first need to handle #65. I want to avoid that contributors start solving quality alerts now, before we change the whole lib, so they don´t wast there time on code that will change anyway.

Beside, we could also replace the current linter with https://github.com/golangci/golangci-lint-action.

**Comment by tannerjones4075** [comment]
I am interested on helping with this issue :)

**Comment by VachaShah** [comment]
Thank you @tannerjones4075! Looking forward to your contributions :)

**Comment by tannerjones4075** [comment]
It seems that this issue should be broken in to several issues: 
1. Define: Code Quality Objectives
2. Code Quality Tool implementation e.g SonarQube
3. Replace/update linter: 
The current version of the linter: https://github.com/opensearch-project/opensearch-go/blob/main/.github/workflows/lint.yml
The version that is currently running is v1.53.3 and the current version is 3.7.0 https://github.com/golangci/golangci-lint-action

I can start with the linter if that is priority.

**Comment by Jakob3xD** [comment]
After defining what Code Quality Objectives we want, we can create sub tasks / issues.

Tow ideas :
- Test Coverage Report that comments to the PR
- govulncheck

> 4\. Replace/update linter:
>         The current version of the linter: https://github.com/opensearch-project/opensearch-go/blob/main/.github/workflows/lint.yml
>         The version that is currently running is v1.53.3 and the current version is 3.7.0 https://github.com/golangci/golangci-lint-action

Version 3.7.0 is the version of the workflow and v1.53.3 is the actual golangci-lint version. The latest version is therefore v1.54.2.

**Comment by tannerjones4075** [comment]
For the SonarQube we could leverage [SonarCloud](https://dev.to/remast/go-for-sonarcloud-with-github-actions-3pmn). This would allows us to create a workflow and implement SonarQube through a GitHub Action.

**Comment by tannerjones4075** [comment]
> For the SonarQube we could leverage [SonarCloud](https://dev.to/remast/go-for-sonarcloud-with-github-actions-3pmn). This would allows us to create a workflow and implement SonarQube through a GitHub Action. For test coverage I came across [this](https://github.com/vladopajic/go-test-coverage). We could have a report generated locally and when a PR is created.
> 
> govulncheck: What do we want the output to be? Do want a report to be [generated](https://github.com/Templum/govulncheck-action/blob/main/README.md)? This will require a token with write access.

It looks like SonarQube has been suggested and research [before](https://github.com/opensearch-project/opensearch-go/pull/313). The conclusion was to update the current golinter, which as been [done](https://github.com/opensearch-project/opensearch-go/pull/394).

**Comment by Jakob3xD** [comment]
Removed govulncheck workflow as it report vulnerabilities of imported libs that we have no control of and can only be solved by updating the go version. As we are a lib and don't build anything, the user need to update their version.
https://github.com/opensearch-project/opensearch-go/pull/421

**Comment by Beeram12** [comment]
> +1 on any kind of static analysis tooling

Hi, I would like to work on this. From the above discussion, I think the objectives of static analysis remain the same, but the `govulncheck` is removed.

**Comment by sean-** [comment]
Looking back at this thread, the actionable items have all landed:

**Linter modernized and expanded** (the path the discussion converged on after #313 ruled out SonarQube in favor of golangci-lint):
- Upgraded to golangci-lint v2.12.2 with the v2 config schema
- 60+ linters enabled (asasalint, bodyclose, contextcheck, errcheck, errorlint, exhaustive, gocritic, gosec, govet, makezero, modernize, nilerr, noctx, nolintlint, prealloc, revive, staticcheck, testifylint, thelper, tparallel, unparam, unused, usestdlibvars, ...)
- Formatters wired in (gci, gofmt, gofumpt, goimports) with local-prefix normalization for `github.com/opensearch-project/opensearch-go`
- Lint runs across three build-tag combinations on every push and PR (`core+plugins`, `+integration`, `+integration+multinode`) so no file escapes coverage due to build tags
- `--fix` enabled in CI to keep the tree drift-free

**Coverage reporting on every PR** (Jakob's follow-up ask in this thread):
- `.codecov.yml` with per-area targets (core, `internal/path/`, v4 `opensearchapi/`, v5preview `opensearchapi/`)
- Codecov uploads from unit, secure-integration, and insecure-integration jobs
- Tracked separately under #157

**govulncheck**: explicitly de-scoped in #421 because, as a library, we can't action upstream Go-stdlib advisories - users update their own toolchain.

**SonarQube specifically**: the thread (#313, then this issue) settled against it in favor of golangci-lint, which is now where ongoing rule additions land.

The "define code quality objectives" sub-item was never converted into concrete acceptance criteria and the linter set is acting as the de facto definition. If a specific quality dimension feels under-checked, the right move is a focused issue ("enable linter X" or "raise rule Y to error"), not a rolling umbrella.

Closing as delivered. Thanks to everyone who weighed in over the years.  Improvements or other forms of contributions are welcome as "code quality" is as much a qualitative objective as much as it is quantitatively enforced via linters, etc (i.e. "patches welcome!").
